### PR TITLE
Movie Generator

### PR DESCRIPTION
**Pull request by Petrychyn**

### CODE REVIEW

File: src/commands/movie.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    }
+    picked
+        .get_mut(&template)
+        .unwrap()
+        .push(String::from(pickedval.clone()));
+    return pickedval;
+}
+
+fn populate_template(tmpl: String, picked: &mut HashMap<String, Vec<String>>) -> String {
+    let mut template = pick(String::from(tmpl), picked);
```
Comment (mp-singh) [inline src/commands/movie.rs]: not sure you need to cast to String::from since tmpl is is already a String

File: src/commands/movie.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+use crate::{MOVIE1, MOVIE_CONTENTS, MOVIE_SYNOPSIS1, MOVIE_SYNOPSIS2};
+
+fn pick(template: String, picked: &mut HashMap<String, Vec<String>>) -> String {
+    let v = MOVIE_CONTENTS.get(&template).unwrap();
+    let r: usize = rand::thread_rng().gen_range(0..v.len());
+    let pickedval = v.get(r).unwrap().clone();
+    if !picked.contains_key(&template) {
+        picked.insert(template.clone(), Vec::new());
+    }
+    picked
```
Comment (mp-singh) [inline src/commands/movie.rs]: isn't picked already mutable? Why are we using get_mut? Am i missing something?

File: src/lib.rs [hunk 3]
```diff
@@ -17,6 +23,36 @@ pub fn count_syllables(word: &str) -> u8 {
...
+                if let Ok(text) = line {
+                    if !text.is_empty() {
+                        cur.push(text);
+                    }
+                }
+            }
+            map.insert(filetype, cur);
+        }
+    }
+    return map;
```
Comment (mp-singh) [inline src/lib.rs]:
```suggestion
    map
```

File: src/commands/movie.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        .get_mut(&template)
+        .unwrap()
+        .push(String::from(pickedval.clone()));
+    return pickedval;
```
Comment (mp-singh) [inline src/commands/movie.rs]:
```suggestion
    pickedval
```

File: src/commands/movie.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    let v = MOVIE_CONTENTS.get(&template).unwrap();
+    let r: usize = rand::thread_rng().gen_range(0..v.len());
+    let pickedval = v.get(r).unwrap().clone();
+    if !picked.contains_key(&template) {
+        picked.insert(template.clone(), Vec::new());
+    }
+    picked
+        .get_mut(&template)
+        .unwrap()
+        .push(String::from(pickedval.clone()));
```
Comment (mp-singh) [inline src/commands/movie.rs]:
```suggestion
        .push(pickedval.clone());
```

File: src/commands/movie.rs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                    if found.is_some() {
+                        return found.unwrap().clone();
+                    }
+                }
+                return pick(tmpl, picked);
+            })
+            .to_mut()
+            .clone();
+    }
+    return template;
```
Comment (mp-singh) [inline src/commands/movie.rs]:
```suggestion
    template
```

File: src/commands/movie.rs (hunk 1 above)
Comment (mp-singh) [inline src/commands/movie.rs]:
```suggestion
    let mut template = pick(tmpl, picked);
```

File: src/commands/movie.rs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    synopsis = MOVIE_SYNOPSIS2
+        .replace_all(&synopsis, |caps: &regex::Captures| {
+            caps.get(1).unwrap().as_str().to_owned() + &caps.get(2).unwrap().as_str().to_uppercase()
+        })
+        .to_mut()
+        .clone();
+    return Movie {
+        title: titlecase(&title),
+        synopsis: synopsis,
+    };
```
Comment (mp-singh) [inline src/commands/movie.rs]:
```suggestion
    Movie {
        title: titlecase(&title),
        synopsis,
    }
```

File: src/lib.rs [hunk 8]
```diff
@@ -17,6 +23,36 @@ pub fn count_syllables(word: &str) -> u8 {
...
+        let curpath = path.unwrap().path();
+        let filename = String::from(curpath.file_name().unwrap().to_str().unwrap());
+        let filetype = String::from(&filename[..(filename.len() - 4)]);
+        if let Ok(lines) = read_lines(curpath) {
+            for line in lines {
+                if let Ok(text) = line {
+                    if !text.is_empty() {
+                        cur.push(text);
+                    }
+                }
```
Comment (mp-singh) [inline src/lib.rs]:
```suggestion
            for line in lines.flatten() {
                if !line.is_empty() {
                    cur.push(line);
                }
            }
```

Comment (mp-singh) [reply src/lib.rs]: https://rust-lang.github.io/rust-clippy/master/index.html#manual_flatten

File: src/lib.rs [hunk 9]
```diff
@@ -17,6 +23,36 @@ pub fn count_syllables(word: &str) -> u8 {
...
 }
 
+pub fn read_dir(dir: String) -> HashMap<String, Vec<String>> {
+    let paths = fs::read_dir(dir).unwrap();
+    let mut map: HashMap<String, Vec<String>> = HashMap::new();
+    for path in paths {
+        let mut cur: Vec<String> = Vec::new();
+        let curpath = path.unwrap().path();
+        let filename = String::from(curpath.file_name().unwrap().to_str().unwrap());
+        let filetype = String::from(&filename[..(filename.len() - 4)]);
```
Comment (mp-singh) [inline src/lib.rs]:
```suggestion
        let filename = curpath.file_name().unwrap().to_str().unwrap().to_string();
        let filetype = filename[..(filename.len() - 4)].to_string();
```

Comment (mp-singh) [reply src/lib.rs]: i just want to keep using to `to_string()` for consistency

File: src/commands/movie.rs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+                    return pick(tmpl, picked);
+                }
+                let num = caps.get(2).unwrap().as_str().parse::<usize>().unwrap() - 1;
+                let list = picked.get(&tmpl);
+                if list.is_some() {
+                    let found = list.unwrap().get(num);
+                    if found.is_some() {
+                        return found.unwrap().clone();
+                    }
+                }
```
Comment (mp-singh) [inline src/commands/movie.rs]:
```suggestion
                if let Some(list) = picked.get(&tmpl) {
                    if let Some(found) = list.get(num) {
                        return found.clone();
                    }
                }
```

**Review by mp-singh — CHANGES_REQUESTED** [review]
Thanks for tackling this. I haven't actually ran it to see the output but i can try it out tomrrow.

File: src/commands/movie.rs (hunk 2 above)
Comment (Petrychyn) [reply src/commands/movie.rs]: Maybe there's no reason. It seems get_mut will return a mutable reference to the value you are getting from the map. Using just get doesn't seem to work.

File: src/commands/movie.rs (hunk 10 above)
Comment (Petrychyn) [reply src/commands/movie.rs]: Changed

File: src/lib.rs (hunk 9 above)
Comment (Petrychyn) [reply src/lib.rs]: Changed

File: src/lib.rs (hunk 8 above)
Comment (Petrychyn) [reply src/lib.rs]: Changed

File: src/commands/movie.rs (hunk 7 above)
Comment (Petrychyn) [reply src/commands/movie.rs]: Changed

File: src/commands/movie.rs (hunk 1 above)
Comment (Petrychyn) [reply src/commands/movie.rs]: Changed

File: src/commands/movie.rs (hunk 6 above)
Comment (Petrychyn) [reply src/commands/movie.rs]: Changed

File: src/commands/movie.rs (hunk 5 above)
Comment (Petrychyn) [reply src/commands/movie.rs]: Changed

File: src/commands/movie.rs (hunk 4 above)
Comment (Petrychyn) [reply src/commands/movie.rs]: Changed

File: src/lib.rs (hunk 3 above)
Comment (Petrychyn) [reply src/lib.rs]: Changed

File: src/commands/movie.rs (hunk 1 above)
Comment (Petrychyn) [reply src/commands/movie.rs]: Changed

File: src/commands/movie.rs (hunk 2 above)
Comment (Petrychyn) [reply src/commands/movie.rs]: Screen Shot 2022-04-23 at 12 18 08 PM

**Review by mp-singh — APPROVED** [review]
Looks good. I still have no idea how this works but it does so boom

**Pull request merged**